### DISCUSSION TITLE
What is the most efficient way to store and edit observable data?

### DISCUSSION THREAD

**Discussion by caiphaav** [mobxjs/mobx#2596] [Q&A] [answered]
I want to store my data in observable collection
But what is the most efficient way to do it?:

- use observable array
- use observable Map
- use observable object

Will observable Map change data faster than array/object like in plain JavaScript?

**Comment by mweststrate** [comment]
Normal data structure complexity rules apply here. If you think it is going
to matter, you should measure your exact use case. It is quite atypical
though if storing data would be the bottleneck in your application.

**Reply by caiphaav** [reply]
The problem is that in my react native app state dynamically change about 50-100 times per second
And thats why i want to find the most performant way - if storing data not effectively it's dropped JS Thread and make awful UX

**Comment by mweststrate** [comment, marked as answer]
That doesn't necessarily sound like a lot. But how do you know storage is
the bottleneck, and not say, updating the UI, or parsing, etc? And why are
you trying to store data faster than could be observed by a human? Sounds
like some batching / debouncing could go a long way as well.

**Reply by caiphaav** [reply]
In my case i have collection with different items. This items separately updates by websocket
Let's say we have list with 100 items, and in one moment of time i can get new data for each item from the server

How I know storage is the bottleneck: 
I write prototype app without UI only with the one screen component and inject mobx store to this component
After model initialization i pressed the button on this screen and make request to websocket and get data from it
After that my JS Thread dropped. But, in fairness, mobx 6 do it more performant than MST or redux in my opinion
JS thread dropped to 45-50 fps only. 
Unfortunately, I can't provide example because of my company rules. But I tried to explain situation in general terms

**Reply by mweststrate** [reply]
Ok, that doesn't sound like we can help, nor that that you know yet where your bottleneck lies, so I recommend to attach a profiler first (if that exists for RN), and otherwise figure out manually what parts of the whole process are slow. Do make sure you use `action` appropriately and consider applying tips written down here: https://mobx.js.org/react-optimizations.html

**Reply by caiphaav** [reply]
Thank you for reply!
